### PR TITLE
enhance virtualenvsetup to not require python-virtualenv installed

### PR DESCRIPTION
**Pull request by tardyp**
depending on python-virtualenv on the host is hard to make work
as the virtualenv version might be different with maybe bad support for https, etc.

This makes the setup dependent on wget and unzip though

I pushed back as much as I could to do this refacto, but I am now convinced this is a better aproach to move forward before switching to a nine metabuildbot.

### CODE REVIEW

**Comment by tardyp** [comment]
I am merging it for the sake of testing on all our weird worker fleet. but I still would like some review

@tomprince @djmitche @sa2ajj

**Comment by tomprince** [comment]
I'm slightly :-1: on this, as I think it makes sense to have some minimal requirements of the workers, and I think virtualenv falls under that heading.

File: test_virtualenv.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from subprocess import call, check_call
+
+from twisted.trial import unittest
+
+from virtualenvsetup import VirtualenvSetup
```
Comment (sa2ajj) [inline test_virtualenv.py]: i'd rather *not* see a relative import

File: virtualenvsetup.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import textwrap
+
+from buildbot import locks
+from buildbot.steps.shell import ShellCommand
+
+# only allow one VirtualenvSetup to run on a slave at a time.  This prevents
+# collisions in the shared package cache.
+veLock = locks.SlaveLock('veLock')
```
Comment (sa2ajj) [inline virtualenvsetup.py]: while we follow Twisted's naming convention, I do not see why we should follow it here.

File: virtualenvsetup.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+
+# we hardcode a decent version of virtualenv here
+VIRTUALENV_URL = 'https://pypi.python.org/packages/bf/f4/43525ce41d4de0128460da3dce059c2c969f4d076bbdbac21b6bc1e996e8/virtualenv-15.0.2-py2.py3-none-any.whl'  # noqa
+VIRTUALENV_FILENAME = VIRTUALENV_URL.split("/")[-1]
+
+
+class VirtualenvSetup(ShellCommand):
+    def __init__(self, virtualenv_dir='sandbox', virtualenv_python='python2.7',
+                 virtualenv_packages=[], **kwargs):
```
Comment (sa2ajj) [inline virtualenvsetup.py]: `[]` is a big **No**

File: virtualenvsetup.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+        echo "Checking for simplejson or json";
+        "$VEPYTHON" -c 'import json' 2>/dev/null || "$VEPYTHON" -c 'import simplejson' ||
+                    "$VE/bin/pip" install simplejson || exit 1;
+        echo "Checking for sqlite3, including pysqlite3 on Python 2.5";
+        "$VEPYTHON" -c 'import sqlite3, sys; assert sys.version_info >= (2,6)' 2>/dev/null ||
+                    "$VEPYTHON" -c 'import pysqlite2.dbapi2' ||
+                    "$VE/bin/pip" install pysqlite || exit 1
+
+        """.format(VIRTUALENV_URL=VIRTUALENV_URL, VIRTUALENV_FILENAME=VIRTUALENV_FILENAME,
```
Comment (sa2ajj) [inline virtualenvsetup.py]: do we really need to have `VIRTUALENV_URL` and `VIRTUALENV_FILENAME`? while lower case versions do not work?

Comment (sa2ajj) [inline virtualenvsetup.py]: just in case, I do not buy "those are corresponding global constants"

**Pull request merged**